### ISSUE TITLE
provide a option to disable shadowcopy in nunit v3

### ISSUE THREAD

**Issue by jmecosta** [nunit/nunit#521]
from vs test runner i get this

NUnit 1.2.0.1 executing tests is EnableShutdown set to True
NUnit 1.2.0.1 executing tests is started
ShadowCopyFiles is set to :False
Loading tests from 
Run started: 
NUnit 1.2.0.1 executing tests is finished

but in command line i dont see any switch to allow me to disable shadowcopy

**Comment by OsirisTerje** [comment]
The shadowcopy in the adapter is controlled by a registry entry, see https://github.com/nunit/nunit-vs-adapter/wiki/Tips-And-Tricks  
On command line the option /noshadow should do the trick

**Comment by jmecosta** [comment]
/noshadow is not availalbe in 3.0 ?

**Comment by jmecosta** [comment]
--noshadow
NUnit Console Runner 3.0.5509
Copyright (C) 2014 Charlie Poole

Invalid argument: --noshadow

NUnit Console Runner 3.0.5509
Copyright (C) 2014 Charlie Poole

Invalid argument: /noshadow

**Comment by CharliePoole** [comment]
Correct, /noshadow was removed from 3.0. See
http://nunit.org/index.php?p=breakingChanges&r=3.0

If you are using the adapter, which works with NUnit 2.6.3, Terje suggested
the (2.6.3) command line. If you are working with 3.0, you shouldn't be
using the adapter at all.

**Comment by CharliePoole** [comment]
BTW, if you have a use case which appears to require disabling shadowcopy,
you should tell us about it on nunit-discuss.

In general, we have found that most such cases presented by users are
actually red herrings and that disabling shadowcopy is almost never
necessary. The only cases I have seen where there is no alternative have to
do with testing assemblies for which you don't have the source code and
therefore cannot fix to work correctly in a shadowcopy situation.  That's
why the option has been removed for the next release.

It would take a very compelling use case to convince me to put it back.

**Comment by jmecosta** [comment]
1000s of legacy tests reading files from working directory. This makes it a
pain to update the test to use 3.0. Having that option available would at
least give me a chance to have a much smoother transition. But anyway I
think am already half way with this for me I think it will not be a big
issue. But I guess there will be plenty of users asking same questions.

Plus as long as the framework allows this I think the test framework should
do the same...

**Comment by CharliePoole** [comment]
So do you want to change this issue to be about NUnit 3.0 dropping /noshadow or do you want to create a new issue? (It's not just a bureaucratic question, since different people work on the two different codebases.)

**Comment by jmecosta** [comment]
Thought this repo was for version 3. I will rephrase the issue so that it
mentions the drop of this option. Please update it if you feel is not
totally correct

**Comment by CharliePoole** [comment]
Yup, this is the right repo. It was the version numbers for the VS adapter
that confused me.

**Comment by rprouse** [comment]
Just back from vacation, so I thought I would jump in here :smile: 

At our company, one of the main reasons we need to turn off shadow copy is because many of our managed assemblies wrap legacy native code. A typical scenario for us is that a C++/CLR assembly links to legacy C++ DLLs (which usually link to other DLLs) and then we use those C++/CLR wrappers from C# code.

In MbUnit, we were never able to get this working with shadow copy enabled because the native assemblies are not copied with shadow copy. We tried several things to get it to work, but the only thing that ever worked was turning off shadow copy. What is the solution for this in NUnit 3 without turning off shadow copy?

Another issue that we will have is that our product is over 6 million lines of code, most of it legacy dating back 30 years. It is not all well written and much of it expects to find resources and other files in directories relative to the executing assembly. We would never get management's approval to fix all of that code just to update to NUnit 3.

**Comment by CharliePoole** [comment]
I'm inclined to put it back in. If we do, then we should have some documentation about how to use it and the drawbacks... mainly that you can't modify the assembly while the gui is open.

**Comment by CharliePoole** [comment]
I have tentatively added this to beta 1 - my initial thought is that it's pretty easy to do. If not, we can push it to a subsequent drop.

However, I'm not sure the proper implementation is to provide a command-line option. I can think of no reason why the console runner needs to use shadowcopy at all. Currently, the DomainManager looks for a package setting of ShadowCopyFiles equal to true or false, with the default being true. We did it this way in 2.x because the gui runner requires it in order to allow recompilation while a test is loaded in the gui.

I can think of no reason why we should ever use shadowcopy for the console. If we support loading from a url or a share in the future, then .NET will automatically use shadowcopy. My understanding is that you can force .NET to use shadowcopy but you can't force it to not use it in situations where it thinks it should.

I think we should experiment with either defaulting to setting it to false in the engine or (even better) not setting it at all if nothing is supplied in the package. That would leave it up to the gui runner to force it to true. There would be no need for a command-line stting in nunit-console.

What think you all?

**Comment by jmecosta** [comment]
Shadow copy can be useful in a console if tests are run in parallel. Our
typical scenario is to have a common test folder were all test assemblies
drop,  we do this to minimize disk space. When running in parallel we will
have copy conflicts if tests run at the same time others projects are
built.

**Comment by jmecosta** [comment]
seems our coverage builds also do not work, not sure if this is related with the fact that shadow copy is enabled by default...

can you point our where can i change the code to disable the shadow copy so i can check that coverage build work...

using opencover

**Comment by CharliePoole** [comment]
Look at DomainManager.cs

**Comment by jmecosta** [comment]
yep now it works fine

**Comment by CharliePoole** [comment]
So... the remaining question for the console runner is this: should it enable or disable shadowcopy by default? If we enable it, we need --noshadow. If we disable it we need --shadow.

**Comment by rprouse** [comment]
I prefer no shadow by default for the console. How about -shadowcopy for
the option?

**Comment by jmecosta** [comment]
I tend to agree that it's good to have them shadow copy as tests will be
written in a more correct way and people will not abuse the getdirectory
method . so keep 2.6.x behavior ?

But each as pros and cons what you decide goes

**Comment by CharliePoole** [comment]
This is an easy fix if we once decide what behavior we want. :smile:

**Comment by finnscho** [comment]
I just trying to upgrade our NUnit from version 2.6 to 3... This is the worst job i ever had... WTF guys? I really don't see any benefit in v3! You disabled features and changed names... That really drives me nuts.. We can't use Nunit v. 3 ... We start to use another testing framework because this has become a pain in the ar...

**Comment by rprouse** [comment]
Other than expected exceptions, and search and replace changes, what issues did you run into? Most people find the upgrade fairly painless, but you might have hit more issues because you waited so long, so some deprecated API's were removed.

**Comment by rprouse** [comment]
You added your comments to an issue on no shadow, is there a problem there? No shadow isn't an option because we don't shadow copy anymore because it causes problems for so many people.
